### PR TITLE
Document TextEdit node

### PR DESCRIPTION
**Pull request by zakstam**
Adds missing documentation for TextEdit Node. Work in progress.

### CODE REVIEW

File: doc/classes/TextEdit.xml [hunk 1]
```diff
@@ -87,6 +87,8 @@
 			<argument index="1" name="adjust_viewport" type="bool" default="true">
 			</argument>
 			<description>
+				Moves the cursor at the specified column index.
```
Comment (StraToN) [inline doc/classes/TextEdit.xml]: "column" should be formatted as "adjust_viewport" below: surround with [code][/code].

File: doc/classes/TextEdit.xml [hunk 2]
```diff
@@ -101,6 +103,8 @@
 			<argument index="3" name="wrap_index" type="int" default="0">
 			</argument>
 			<description>
+				Moves the cursor at the specified line index.
+				If [code]adjust_viewport[/code] is set to true, the viewport will center at the cursor position after the move occurs.
```
Comment (StraToN) [inline doc/classes/TextEdit.xml]: Same as above: 
"line" should be formatted as "adjust_viewport" below: surround with [code][/code].
Also: 2 parameters descriptions are missing: can_be_hidden and wrap_index

File: doc/classes/TextEdit.xml [hunk 3]
```diff
@@ -146,6 +150,7 @@
 			<argument index="0" name="keyword" type="String">
 			</argument>
 			<description>
+				Returns the [Color] of a keyword at the specified index.
```
Comment (StraToN) [inline doc/classes/TextEdit.xml]: This parameter is not an index, but a keyword. Should be:
Returns the [Color] of the specified [code]keyword[/code].

File: doc/classes/TextEdit.xml [hunk 4]
```diff
@@ -218,6 +224,7 @@
 			<argument index="0" name="keyword" type="String">
 			</argument>
 			<description>
+				Returns whether the keyword of the specified index has a color set to it or not.
```
Comment (StraToN) [inline doc/classes/TextEdit.xml]: No index here either.

**Review by StraToN — CHANGES_REQUESTED** [review]

File: doc/classes/TextEdit.xml (hunk 3 above)
Comment (zaksnet) [reply doc/classes/TextEdit.xml]: I'm confuzed, what is the index argument for then?

Comment (zaksnet) [reply doc/classes/TextEdit.xml]: You are right that its a sting keyword by looking at the code
`Color TextEdit::get_keyword_color(String p_keyword) const {`
but why is there an index argument?

Comment (StraToN) [reply doc/classes/TextEdit.xml]: There is no index argument provided to the function. The `index` attribute in the XML node `<argument>` is simply the index of the argument itself, as an information for the documentation parser. 
If `<argument index="0">` then this argument is the first parameter of the function.
If `<argument index="1">` then this argument is the second parameter of the function, etc.

Comment (zaksnet) [reply doc/classes/TextEdit.xml]: Sorry my fault 😳

File: doc/classes/TextEdit.xml (hunk 2 above)
Comment (zaksnet) [reply doc/classes/TextEdit.xml]: I'm not sure what wrap_index is.

File: doc/classes/TextEdit.xml [hunk 5]
```diff
@@ -490,8 +506,10 @@
 			Undoes the previous action.
 		</constant>
 		<constant name="MENU_REDO" value="6" enum="MenuItems">
+			Redoes the previous action.
 		</constant>
 		<constant name="MENU_MAX" value="7" enum="MenuItems">
+
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: The description should not be empty ;)
You can use something like "Represents the size of the [enum MenuItems] enum."

File: doc/classes/TextEdit.xml [hunk 6]
```diff
@@ -259,6 +268,7 @@
 			<argument index="0" name="option" type="int">
 			</argument>
 			<description>
+				Triggers a right click menu action by the specified index. See [MENU_*] enums for a list of available indexes.
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: `[MENU_*]` will not work. It should be "See the [enum MenuItems] enum values for...".

File: doc/classes/TextEdit.xml (hunk 2 above)
Comment (StraToN) [reply doc/classes/TextEdit.xml]: I guess this is the column number where the TextEdit will wrap to a new line. That's useful for source code if you don't want a line of code to have length bigger than `wrap_index`.
**This is just a guess, it probably requires a test first.**

File: doc/classes/TextEdit.xml [hunk 7]
```diff
@@ -87,6 +87,8 @@
 			<argument index="1" name="adjust_viewport" type="bool" default="true">
 			</argument>
 			<description>
+				Moves the cursor at the specified [code]column[/code] index.
+				If [code]adjust_viewport[/code] is set to true, the viewport will center at the cursor position after the move occurs. Default value is [code]true[/code]
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: Missing dot at the end of the line.

File: doc/classes/TextEdit.xml [hunk 8]
```diff
@@ -101,6 +103,9 @@
 			<argument index="3" name="wrap_index" type="int" default="0">
 			</argument>
 			<description>
+				Moves the cursor at the specified [code]line[/code] index.
+				If [code]adjust_viewport[/code] is set to true, the viewport will center at the cursor position after the move occurs. Default value is [code]true[/code].
+				If [code]can_be_hidden[/code] is set to true, the specified [code]line[/code] can be hidden using [member set_line_as_hidden]. Default value is [code]true[/code]
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: Missing dot at the end of the line.

File: doc/classes/TextEdit.xml [hunk 9]
```diff
@@ -226,7 +234,7 @@
 			<argument index="0" name="text" type="String">
 			</argument>
 			<description>
-				Insert a given text at the cursor position.
+				Insert a the specified text at the cursor position.
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: "a" should be removed.

File: doc/classes/TextEdit.xml [hunk 10]
```diff
@@ -235,7 +243,7 @@
 			<argument index="0" name="line" type="int">
 			</argument>
 			<description>
-				Returns if the given line is folded.
+				Returns if the line at the specified index is folded or not.
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: "Returns whether" would be better.

Comment (akien-mga) [reply doc/classes/TextEdit.xml]: BTW, I wonder if it's better to use "or not" in "Returns whether ... or not" or if it should be omitted.
CC @cbscribe @NathanLovato. This should likely go in the docs writing guidelines too as "Returns whether ..." should likely be the recommended construct for any boolean getter.

File: doc/classes/TextEdit.xml [hunk 11]
```diff
@@ -259,6 +268,7 @@
 			<argument index="0" name="option" type="int">
 			</argument>
 			<description>
+				Triggers a right click menu action by the specified index. See [enum MenuItems] for a list of available indexes.
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: We seem to use both "indexes" and "indices" throughout the classref, we should probably decide which one is correct and enforce its use. CC @cbscribe @NathanLovato

File: doc/classes/TextEdit.xml [hunk 12]
```diff
@@ -407,8 +421,10 @@
 			If [code]true[/code], line numbers are displayed to the left of the text.
 		</member>
 		<member name="smooth_scrolling" type="bool" setter="set_smooth_scroll_enable" getter="is_smooth_scroll_enabled">
+			If [code]true[/code], sets the [code]step[/code] of the scrollbars to [code]0.25[/code] which occurs in smoother scrolling.
```
Comment (akien-mga) [inline doc/classes/TextEdit.xml]: I'm not sure "occurs in" is correct here. "results in" is probably better, and easier to understand for non native speakers.

**Comment by zaksnet** [comment]
Some of the function/member/signal names are misleading while I feel they should be self-explaining. Would it be ok if I changed their names? I suspect these would be breaking changes.

**Comment by akien-mga** [comment]
That would break compatibility yes, so we can't change them now.

You can list them in #16863 for 4.0.

**Comment by zaksnet** [comment]
How about undocumented functions for this node? For example, i can see there is a theme option for "member_variable_color" but "add_member_keyword", "has_member_color" etc are missing from the docs (while they exist in item_list.cpp). Is it ok if i document these?

File: doc/classes/TextEdit.xml (hunk 10 above)
Comment (zaksnet) [reply doc/classes/TextEdit.xml]: That could also be used for setters not just getters. Example:
`<method name="set_line_as_hidden">`
"If [code]true[/code], hides the line of the specified index." 
Could be:
"Determines whether the line at the specified index will be hidden or not."

**Comment by mrcdk** [comment]
Oh, I didn't notice this PR and also documented some of the methods and properties here https://github.com/godotengine/godot/pull/29823 I see that the documentation here is better so would you rather get this PR merged and I'll make the changes to my PR or get my PR merged and you make changes to this PR?

**Comment by zaksnet** [comment]
> Oh, I didn't notice this PR and also documented some of the methods and properties here #29823 I see that the documentation here is better so would you rather get this PR merged and I'll make the changes to my PR or get my PR merged and you make changes to this PR?

Sorry, I'm new with GIT and I'm not sure how to merge. You've clearly put a lot more time into this than me, into this so i think we should merge mine with yours.

**Comment by mrcdk** [comment]
Okay, I'll wait and update my PR whenever this PR gets merged 😀

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.1.2.

**Pull request merged**